Source/Mothership/AppTransitions.cpp

```diff
@@ -0,0 +1,195 @@
+ *
+ *  2. starts all of the threads for each core.
+ *
+ *  3. triggers backend-level application execution for each core.
+ *
+ *  4. initialises the supervisor device on this Mothership.
+ *
+ * After 3, if all is well, each thread sends a BARRIER packet (Tinsel message)
+ * to the Mothership (which is handled by the BackendInputBroker thread). Note
+ * that steps 2 and 3 are handled in different loops. This is because MFN
+ * stated (at some arbitrary point in the past) that starting all of the
+ * threads, then kicking them off is safer than kicking off some cores before
+ * other cores' threads have started. */
+void Mothership::initialise_application(AppInfo* app)
+{
+    std::map<uint32_t, CoreInfo>::iterator coreIt;
+
+    /* Staging variables for address components (for Tinsel speak) */
+    uint32_t meshX;
+    uint32_t meshY;
+    uint32_t coreId;
+    uint32_t threadId;  /* Not used, as far as I can see. */
+
+    /* Modes to reduce code repetition between steps 2 and 3. */
+    bool mode;
+
+    debug_post(590, 1, app->name.c_str());
+
+    app->state = LOADING;  /* 0: Set the application state to LOADING, duh. */
+
+    /* 1: For each core, call backend->loadInstrsOntoCore and
+     * loadDataViaCore. */
+    pthread_mutex_lock(&(threading.mutex_backend_api));
+    for (coreIt = app->coreInfos.begin(); coreIt != app->coreInfos.end();
+         coreIt++)
+    {
+        backend->fromAddr(coreIt->first, &meshX, &meshY, &coreId, &threadId);
+        debug_post(578, 4, coreIt->second.codePath.c_str(),
+                   hex2str(meshX).c_str(), hex2str(meshY).c_str(),
+                   hex2str(coreId).c_str());
+        backend->loadInstrsOntoCore(coreIt->second.codePath.c_str(),
+                                   meshX, meshY, coreId);
+        debug_post(577, 4, coreIt->second.dataPath.c_str(),
+                   hex2str(meshX).c_str(), hex2str(meshY).c_str(),
+                   hex2str(coreId).c_str());
+        backend->loadDataViaCore(coreIt->second.dataPath.c_str(),
+                                 meshX, meshY, coreId);
+    }
+
+    /* 2: For each core, kick off the threads (mode=false).
+     * 3: For each core, start execution (mode=true). */
+    mode = false;
+    do
+    {
+        for (coreIt = app->coreInfos.begin(); coreIt != app->coreInfos.end();
+             coreIt++)
+        {
+            backend->fromAddr(coreIt->first, &meshX, &meshY, &coreId,
+                              &threadId);
+
+            if (!mode)  /* 2 */
+            {
+                debug_post(
+                    580, 4, hex2str(meshX).c_str(), hex2str(meshY).c_str(),
+                    hex2str(coreId).c_str(),
+                    uint2str(coreIt->second.threadsExpected.size()).c_str());
+                /* Note that startOne can hang for the Tinsel backend if the
+                 * number of threads expected is greater than the number of
+                 * threads that the core will start - this is because startOne
```

Comment:
I put this comment here because I had a nasty time debugging this issue when I introduced such a bug. It's a pain, because one might make a small change to the binary build process, and the only behaviour they observe is a hang precisely here (which is hard to identify, because the Mothership might not be easily debug-able if it is running on a different box).

Assuming the binary generation code is written correctly, and assuming the message mechanism is also written correctly, we would never observe this behaviour.